src/models/geograph.py

```diff
@@ -66,50 +305,135 @@ def _dataframe_to_graph(
         if attributes is None:
             attributes = df.columns.tolist()
+        elif not set(attributes).issubset(set(df.columns)):
+            raise ValueError("`attributes` must only contain column names in `df`.")
 
-        geom = df.geometry.tolist()
+        geom: Dict[int, shapely.Polygon] = df.geometry.to_dict()
         # Build dict mapping hashable unique object ids for each polygon
         # to their index in geom
-        id_dict = {id(polygon): index for index, polygon in enumerate(geom)}
+        id_dict: Dict[int, int] = {
+            id(polygon): index for index, polygon in geom.items()
+        }
         # Build Rtree from geometry
-        tree = STRtree(geom)
+        tree = STRtree(geom.values())
         # this dict maps polygon indices in df to a list
         # of neighbouring polygon indices
         graph_dict = {}
 
         # Creating nodes (=vertices) and finding neighbors
         for index, polygon in tqdm(
-            enumerate(geom),
+            geom.items(),
             desc="Step 1 of 2: Creating nodes and finding neighbours",
             total=len(geom),
         ):
-            # find the indexes of all polygons which touch the borders of or
-            # overlap with this one
-            neighbours = [
-                id_dict[id(nbr)]
-                for nbr in tree.query(polygon)
-                if nbr.touches(polygon) or nbr.overlaps(polygon)
-            ]
+            if tolerance > 0:
+                # Expand the borders of the polygon by `tolerance```
+                new_polygon = polygon.buffer(tolerance)
+                # find the indexes of all polygons which intersect with this one
+                neighbours: List[int] = [
+                    id_dict[id(nbr)]
+                    for nbr in tree.query(new_polygon)
+                    if nbr.intersects(new_polygon)
+                    and id_dict[id(nbr)] != id_dict[id(polygon)]
+                ]
+            else:
+                neighbours = [
+                    id_dict[id(nbr)]
+                    for nbr in tree.query(polygon)
+                    if nbr.intersects(polygon)
+                    and id_dict[id(nbr)] != id_dict[id(polygon)]
+                ]
             # this dict maps polygon indices in df to a list
             # of neighbouring polygon indices
             graph_dict[index] = neighbours
             row = df.loc[index]
             # getting dict of column values in row
             row_attributes = dict(zip(attributes, [row[attr] for attr in attributes]))
             # add each polygon as a node to the graph with all attributes
-            rep_point = row.geometry.representative_point()
-            self.graph.add_node(index, rep_point=rep_point, **row_attributes)
+            self.graph.add_node(
+                index,
+                rep_point=polygon.representative_point(),
+                area=polygon.area,
+                perimeter=polygon.length,
+                **row_attributes,
```

Comment:
Another thought that just crossed my mind: 
- When we load from raster data, the class value will always be called `raster_val` at the moment right? (b.c. that's the default in your polygonize above). 
- When we load from a dataframe, the class value will always be whatever it's called in that df

==> Do we want to harmonise this in some way? E.g. a unified name like `node_value` or your `class_label` from below or so, such that we can later easily access this in queries over the graph?

Reply:
If you check the polygonise function I actually changed `raster_val` to `class_label` - as I mentioned in the PR I would like to harmonise on `class label` - currently the only way it can be anything different is if a different named column is passed in a dataframe, but I'm not sure how to prevent that.

Reply:
If they're passing in a dataframe then they might as well just rename it to `class_label` anyway, right? Maybe we should just throw an error if it doesn't include `class_label`.

Reply:
Agreed, that would simplify things a lot

Reply:
@Croydon-Brixton I added this as a check to the `_dataframe_to_graph()` function, but I realised that if we load in data from a vector file, such as a shapefile, this will automatically trigger the check since the resulting dataframe will not have a class label column (or it will be named something different).

Therefore I have added a new function in `vector_utils.py` which loads in any vector data, renames a specified column to "class_label", saves it in the GPKG format and returns the dataframe.